pkg/oc/cli/cmd/login/loginoptions_test.go

```diff
@@ -256,6 +266,96 @@ func TestDialToHTTPServer(t *testing.T) {
 	}
 }
 
+type oauthMetadataResponse struct {
+	metadata *util.OauthAuthorizationServerMetadata
+}
+
+func (r *oauthMetadataResponse) Serialize() string {
+	b, err := json.Marshal(r.metadata)
+	if err != nil {
+		return ""
+	}
+
+	return string(b)
+}
+
+func TestPreserveErrTypeAuthInfo(t *testing.T) {
+	invoked := make(chan struct{}, 2)
+	metadataResponse := &oauthMetadataResponse{}
+
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		invoked <- struct{}{}
+
+		if r.URL.Path == tokencmd.OauthMetadataEndpoint {
+			w.WriteHeader(http.StatusOK)
+			io.WriteString(w, metadataResponse.Serialize())
+			return
+		}
+		w.WriteHeader(http.StatusUnauthorized)
+	}))
+
+	metadataResponse.metadata = &util.OauthAuthorizationServerMetadata{
+		Issuer:                        server.URL,
+		AuthorizationEndpoint:         server.URL + "/oauth/authorize",
+		TokenEndpoint:                 server.URL + "/oauth/token",
+		CodeChallengeMethodsSupported: []string{"plain", "S256"},
+	}
+	defer server.Close()
+
+	clientConfig := &restclient.Config{
+		Host: server.URL,
+	}
+
+	options := newLoginOptions(server.URL, "test", "test", false)
+	options.Config = clientConfig
+
+	err := options.gatherAuthInfo()
+	if err == nil {
+		t.Fatalf("expecting unauthorized error when gathering authinfo")
+	}
+
+	if !kapierrs.IsUnauthorized(err) {
+		t.Fatalf("expecting error of type metav1.StatusReasonUnauthorized, but got %v", reflect.TypeOf(err))
+	}
+}
+
+func newLoginOptions(server string, username string, password string, insecure bool) *LoginOptions {
+	flagset := pflag.NewFlagSet("test-flags", pflag.ContinueOnError)
+	flags := []string{}
+	clientConfig := defaultClientConfig(flagset)
+	flagset.Parse(flags)
+
+	startingConfig, _ := clientConfig.RawConfig()
+
+	loginOptions := &LoginOptions{
+		Server:             server,
+		StartingKubeConfig: &startingConfig,
+		Username:           username,
+		Password:           password,
+		InsecureTLS:        insecure,
+
+		Out:    ioutil.Discard,
+		ErrOut: ioutil.Discard,
+	}
+
+	return loginOptions
+}
+
+func defaultClientConfig(flags *pflag.FlagSet) kclientcmd.ClientConfig {
```

Comment:
Flags, wat?

Reply:
Heh, had originally grabbed these two helpers [from here](https://github.com/openshift/origin/blob/master/test/integration/login_test.go#L127). Went ahead and cleaned the test up - no longer need them for this test